### PR TITLE
[MemDepAnalysis] Don't reuse NonLocalPointerDeps cache if memory location size differs

### LINKED ISSUE #111585: Miscompile in GVN due to MemDepAnalysis caching
The following miscompile was uncovered in GVN, to reproduce:
`opt --passes=gvn  gvntest.ll` on [gvntest.ll.txt](https://github.com/user-attachments/files/17299256/gvntest.ll.txt)

The miscompiled output is: 
[testafterbad.ll.txt](https://github.com/user-attachments/files/17299260/testafterbad.ll.txt)

It appears that if the cache entry is cleared at MemoryDependenceAnalysis.cpp:1060 with `NonLocalPointerDeps.erase(CacheKey);` before the insert, the test is correct.
The resulting IR with this change: 
[testaftergood.ll.txt](https://github.com/user-attachments/files/17299268/testaftergood.ll.txt)

### PR DESCRIPTION
**Pull request by aeubanks**
As seen in #111585, we can end up using a previous cache entry where the size was too large and was UB.

Compile time impact: https://llvm-compile-time-tracker.com/compare.php?from=6a863f7e2679a60f2f38ae6a920d0b6e1a2c1690&to=faccf4e1f47fcd5360a438de2a56d02b770ad498&stat=instructions:u.

Fixes #111585.

### CODE REVIEW

**Comment by aeubanks** [comment]
Still trying to understand the code to see if this is the correct fix, but figured I'd send this out to get thoughts

I'm having trouble coming up with a test case that isn't brittle.

**Review by nikic — COMMENTED** [review]
Can you please update the failing Transforms/GVN/PRE/rle.ll test?

**Review by nikic — COMMENTED** [review]
I think that in principle, not reusing cache entries from a greater size for a smaller one is correct. As you point out, it may be incorrect if the larger size is UB.

It should still be fine to reuse a cache entry without size for one with, and it would also be valid to use an upperBound() larger size instead of a precise() one. But possibly it's not worth bothering with it.

**Comment by aeubanks** [comment]
updated test

File: llvm/test/Transforms/GVN/PRE/rle.ll [hunk 1]
```diff
@@ -98,12 +98,13 @@ define ptr @coerce_mustalias3(float %V, ptr %P) {
...
 ; CHECK-NEXT:    [[A:%.*]] = load i32, ptr [[P:%.*]], align 4
-; CHECK-NEXT:    [[TMP1:%.*]] = bitcast i32 [[A]] to float
+; CHECK-NEXT:    [[TMP1:%.*]] = load float, ptr [[P]], align 4
 ; CHECK-NEXT:    br i1 [[COND:%.*]], label [[T:%.*]], label [[F:%.*]]
 ; CHECK:       T:
 ; CHECK-NEXT:    ret float [[TMP1]]
 ; CHECK:       F:
-; CHECK-NEXT:    ret float [[TMP1]]
+; CHECK-NEXT:    [[X:%.*]] = bitcast i32 [[A]] to float
+; CHECK-NEXT:    ret float [[X]]
```
Comment (nikic) [inline llvm/test/Transforms/GVN/PRE/rle.ll:107]: Why does this fail to fold now? There are no access size changes going on here.

Comment (aeubanks) [reply llvm/test/Transforms/GVN/PRE/rle.ll]: sorry, I had messed up which version of opt I had used for UTC

**Review by nikic — COMMENTED** [review]
This looks good to me, but I'd still like to have a test case, even if it's brittle.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by Disservin** [comment]
@aeubanks not sure if this got lost but should this #80494 be added as a test ?

**Pull request merged**